src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,519 @@
+                true
+            } else {
+                // Different tables, choose `encoded_actual_end_table_prefix` as a split key.
+                // Note that table id does not grow by 1, so have to use
+                // `encoded_actual_end_table_prefix`.
+                // See more: https://github.com/pingcap/tidb/issues/4727
+                status.last_encoded_table_prefix = Some(encoded_actual_end_table_prefix.to_vec());
+                false
+            }
+        }
+        // The region starts from tabel area to non-table area.
+        (Ordering::Equal, Ordering::Greater) => {
+            // As the comment above, outside needs scan for finding a split key.
+            status.first_encoded_table_prefix = encoded_actual_start_table_prefix.to_vec();
+            false
+        }
+        _ => panic!(
+            "start_key {:?} and end_key {:?} out of order",
+            actual_start_key,
+            actual_end_key
+        ),
+    }
+}
+
+/// Feed keys in order to find the split key.
+fn check_key(status: &mut TableStatus, key: &[u8]) -> Option<Vec<u8>> {
+    if let Some(last_encoded_table_prefix) = status.last_encoded_table_prefix.take() {
+        // `before_check` found a split key.
+        Some(keys::data_key(&last_encoded_table_prefix))
+    } else if let Some(current_encoded_table_prefix) =
+        cross_table(&status.first_encoded_table_prefix, key)
+    {
+        Some(keys::data_key(&current_encoded_table_prefix))
+    } else {
+        None
+    }
+}
+
+/// If `current_data_key` does not belong to `encoded_table_prefix`.
+/// it returns the `current_data_key`'s encoded table prefix.
+fn cross_table(encoded_table_prefix: &[u8], current_data_key: &[u8]) -> Option<Vec<u8>> {
+    if !keys::validate_data_key(current_data_key) {
+        return None;
+    }
+    let current_encoded_key = keys::origin_key(current_data_key);
+    if !current_encoded_key.starts_with(table_codec::TABLE_PREFIX) {
+        return None;
+    }
+
+    let current_encoded_table_prefix =
+        if let Some(key) = extract_encoded_table_prefix(current_encoded_key) {
+            key
+        } else {
+            return None;
+        };
+
+    if encoded_table_prefix != current_encoded_table_prefix {
+        Some(current_encoded_table_prefix.to_vec())
+    } else {
+        None
+    }
+}
+
+#[allow(collapsible_if)]
+fn bound_keys(db: &DB, region: &Region) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+    let start_key = keys::enc_start_key(region);
+    let end_key = keys::enc_end_key(region);
+    let mut first_key = None;
+    let mut last_key = None;
+
+    let iter_opt = IterOption::new(Some(end_key), false);
+    let mut iter = box_try!(db.new_iterator_cf(CF_WRITE, iter_opt));
+
+    // the first key
+    if iter.seek(start_key.as_slice().into()) {
+        let key = iter.key().to_vec();
+        first_key = Some(key);
+    } // else { No data in this CF }
+
+    // the last key
+    if iter.seek(SeekKey::End) {
```

Comment:
I think it should be `seek_for_prev`. And seems no case is covering it.

Reply:
The iter has an upper_bound. It is covered at https://github.com/pingcap/tikv/pull/2378/files/0594bed3125442cb487eeeb47d2f7cd0c6a8f0e7#diff-4c9d2f3441c437af3def4963493f90d9R328